src/git-unix/ogit-http-fetch/main.ml

```diff
@@ -0,0 +1,169 @@
+    level (Fmt.option Fmt.string)
+    (Option.map (pad 10) header)
+
+let reporter ppf =
+  let report src level ~over k msgf =
+    let k _ =
+      over ();
+      k ()
+    in
+    let with_src_and_stamp h _ k fmt =
+      let dt = Mtime.Span.to_us (Mtime_clock.elapsed ()) in
+      Fmt.kpf k ppf
+        ("%s %a %a: @[" ^^ fmt ^^ "@]@.")
+        (pad 10 (Fmt.str "%+04.0fus" dt))
+        pp_header (level, h)
+        Fmt.(styled `Magenta string)
+        (pad 10 @@ Logs.Src.name src)
+    in
+    msgf @@ fun ?header ?tags fmt -> with_src_and_stamp header tags k fmt
+  in
+  { Logs.report }
+
+let setup_logs style_renderer level ppf =
+  Fmt_tty.setup_std_outputs ?style_renderer ();
+  Logs.set_level level;
+  Logs.set_reporter (reporter ppf);
+  let quiet = match style_renderer with Some _ -> true | None -> false in
+  quiet, ppf
+
+type error = [ `Store of Store.error | `Sync of Sync.error ]
+
+let store_err err = `Store err
+let sync_err err = `Sync err
+
+let pp_error ppf = function
+  | `Store err -> Fmt.pf ppf "(`Store %a)" Store.pp_error err
+  | `Sync err -> Fmt.pf ppf "(`Sync %a)" Sync.pp_error err
+
+let main (references : (Git.Reference.t * Git.Reference.t) list)
+    (directory : string option) (repository : Uri.t) =
+  let ( >>?= ) = Lwt_result.bind in
+  let ( >>!= ) v f = Lwt_result.map_err f v in
+  let root : Fpath.t =
+    Fpath.v @@ match directory with Some p -> p | None -> Sys.getcwd ()
+  in
+  let resolvers : Conduit.resolvers =
+    Conduit_lwt.empty
+    |> Conduit_lwt.add Conduit_lwt.TCP.protocol
+         (Conduit_lwt.TCP.resolve ~port:80)
+  in
+  let remote_references = List.map fst references in
+  Lwt.return @@ Smart_git.endpoint_of_string (Uri.to_string repository)
+  >>!= store_err
+  >>?= fun endpoint ->
+  Store.v root >>!= store_err >>?= fun store ->
+  Sync.fetch ~resolvers endpoint store (`Some remote_references) >>!= sync_err
+
+open Cmdliner
+
+module Flag = struct
+  let output_value =
+    let parse str =
+      match str with
+      | "stdout" -> Ok Fmt.stdout
+      | "stderr" -> Ok Fmt.stderr
+      | s -> Error (`Msg (Fmt.strf "%s is not an output." s))
+    in
+    let print ppf v =
+      Fmt.pf ppf "%s" (if v == Fmt.stdout then "stdout" else "stderr")
+    in
+    Arg.conv ~docv:"<output>" (parse, print)
+
+  let output =
+    let doc = "Output of the progress status" in
+    Arg.(
+      value
+      & opt output_value Fmt.stdout
+      & info [ "output" ] ~doc ~docv:"<output>")
+
+  let progress =
+    let doc =
+      "Progress status is reported on the standard error stream by default \
+       when it is attached to a terminal, unless -q is specified. This flag \
+       forces progress status even if the standard error stream is not \
+       directed to a terminal."
+    in
+    Arg.(value & flag & info [ "progress" ] ~doc)
+
+  let uri =
+    let parse str = Ok (Uri.of_string str) in
+    let print = Uri.pp_hum in
+    Arg.conv ~docv:"<uri>" (parse, print)
```

Comment:
I think, instead to use `Uri.of_string` in this argument, you can use `Smart_git.endpoint_of_string` directly here. By this way, `repository` will be a correct value in `main`.

However, if we go to this way, we extend the ability of `Sync` to `fetch`/`clone` over:
- HTTP / HTTPS
- SSH
- TCP

And `ogit-http-fetch` will be more `ogit-fetch` then. To handle these protocols, you should fill `resolvers` with an SSH implementation and an other TCP implementation on the 9418 port (the port used when we clone with `git://`).

Reply:
NOTE: at this level, it's fine to require `awa-ssh`/`cohttp-lwt-unix` and a TLS stack to be able to handle HTTPS and SSH.